### PR TITLE
Stats summary & operations support

### PR DESCRIPTION
**Pull request by asilvas**
Two objectives with this PR:

* Provide an approximation for operation statistics
* Expose an accumulation helper for surfacing aggregated stats over across time and tensors

### CODE REVIEW

File: scripts/gen_binding.py [hunk 1]
```diff
@@ -523,12 +523,15 @@ def normalize_arg(arg):
     const [t0, b0] = recorded_stat
     const dt = performance.now() - t0
     const db = fl.bytesUsed.native() - b0
+    const ops = BigInt([{','.join(js_tensor_args)}].reduce((ops, t) => ops + t.elements, 0))
```
Comment (asilvas) [inline scripts/gen_binding.py:526]: Not all ops are created equal so there is probably a more intelligent way to track "useful" ops, but hopefully this is a sufficient approximation for now

Comment (bwasti) [reply scripts/gen_binding.py:526]: is this an approximation of `flops`?  if so, I think we should jump into immediately making a registry to calculate it exactly for each operation.  most are going to be point-wise (as you have here), but the ones that really count will not be (matmul, conv)

File: shumai/tensor/stats.ts [hunk 2]
```diff
@@ -22,12 +22,86 @@ export function collectStats(tensors: Tensor[]) {
...
     }
   }
   return stats
 }
+
+export type StatsEntry = {
+  count: number
+  time: number
+  bytes: bigint
+  ops: bigint
```
Comment (bwasti) [inline shumai/tensor/stats.ts:41]: let's rename this `flops` everywhere.  Admittedly they're not all floating point, but the nomenclature is common in the literature and it'll prevent confusion with `op` as in "operation"

File: shumai/tensor/stats.ts [hunk 3]
```diff
@@ -22,12 +22,86 @@ export function collectStats(tensors: Tensor[]) {
...
+    }
+  }
+
+  for (const t of tensors) {
+    if (!t.stats) {
+      throw new Error(`tensor has no stats ${t.ptr}:${t.shape}`)
+    }
+
+    for (const key of Object.keys(t.stats)) {
+      const op = /^(\n)?(.*)@/.exec(key)?.[2] || key
```
Comment (bwasti) [inline shumai/tensor/stats.ts:69]: why not `t.op`? perhaps I don't understand this line

File: shumai/tensor/stats.ts [hunk 4]
```diff
@@ -22,12 +22,86 @@ export function collectStats(tensors: Tensor[]) {
...
+      const entry = t.stats[key]
+      stats.totals.count += entry.count
+      stats.totals.time += entry.time
+      stats.totals.bytes += entry.bytes
+      stats.totals.ops += entry.ops
+
+      const stackEntry = stats.entriesByStack.get(key)
```
Comment (bwasti) [inline shumai/tensor/stats.ts:76]: we'll probably want to refactor the key to be a short hash (as in https://github.com/facebookresearch/shumai/blob/main/shumai/tensor/tensor.ts#L428) and carry around the actual stack string separately.  Not necessary for this PR, but just a thought

**Comment by bwasti** [comment]
Overall, I like this!

The one thing I'd request is factoring out the logic for calculating `ops` (renaming it to `flops` for clarity) and creating a registry (like register_gradients.ts) for each operator.  No need to populate the registry in this PR (I can help out with that), but it'll be good to collect real FLOPs formulas for operators

**Review by bwasti — CHANGES_REQUESTED** [review]

File: shumai/tensor/stats.ts (hunk 4 above)
Comment (asilvas) [reply shumai/tensor/stats.ts:76]: Not sure your motive on this one. Strings are immutable and the length has no impact on the key lookup. Computing a hash is more intensive.

File: scripts/gen_binding.py (hunk 1 above)
Comment (asilvas) [reply scripts/gen_binding.py:526]: I'm probably not the best person to fill out that registry as I'm not certain the exact complexity of every method, but I can give it a shot.

File: shumai/tensor/stats.ts (hunk 4 above)
Comment (bwasti) [reply shumai/tensor/stats.ts:76]: oh good point re perf, I didn't realize even long strings would be interned in JS.  I'm personally biased towards readable keys (the stacks are unwieldy to print)

File: shumai/tensor/stats.ts (hunk 3 above)
Comment (asilvas) [reply shumai/tensor/stats.ts:69]: Tensors track stats across tensor operations so we must pluck the actual operation from the stack trace.

File: shumai/tensor/op_to_flops.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+export function opToFlops(op: string, args: any[]): bigint {
+  switch (op) {
+    case 'matmul':
+      return BigInt(Math.round(Math.pow(args[0].elements, 2.37188)))
```
Comment (asilvas) [inline shumai/tensor/op_to_flops.ts:4]: Doesn't account for scalars but wasn't sure if there was a more accurate calculation.

**Comment by asilvas** [comment]
Will reopen after fixes and improvements

File: shumai/tensor/op_to_flops.ts (hunk 5 above)
Comment (bwasti) [reply shumai/tensor/op_to_flops.ts]: you'll want to do something like `args[0].shape[0] * args[1].shape[0] * args[1].shape[1] * 2`, but this is fine to check in for now

File: scripts/gen_binding.py [hunk 6]
```diff
@@ -508,6 +508,7 @@ def normalize_arg(arg):
   const requires_stats = {js_requires_stats}
 
   let stats = null
```
Comment (bwasti) [inline scripts/gen_binding.py:510]: this remains unchanged right? I guess we can inline that into `t.stats = null` later in the code

Comment (asilvas) [reply scripts/gen_binding.py:510]: I kept the core stats logic as untouched as possible to keep scope down, but have ideas for cleanup/enhancements in follow up.

**Pull request merged**